src/main/scala/org/constellation/primitives/storage/CheckpointService.scala

```diff
@@ -54,23 +59,104 @@ class CheckpointBlocksMemPool(size: Int = 50000)(implicit dao: DAO)
 
 object CheckpointService {
   def apply(implicit dao: DAO, size: Int = 50000) = new CheckpointService(dao, size)
+
+  def convert(merkle: CheckpointCacheDataMerkle)(implicit dao: DAO): CheckpointCacheData = {
+    val txs = fetchTransactions(merkle.checkpointBlock.transactionsMerkleRoot)
+    val msgs =
+      merkle.checkpointBlock.messagesMerkleRoot
+        .fold(Seq[ChannelMessage]())(mr => fetchMessages(mr))
+    val notifications =
+      merkle.checkpointBlock.notificationsMerkleRoot.fold(Seq[PeerNotification]())(mr => fetchNotifications(mr))
+    CheckpointCacheData(
+      Some(
+        CheckpointBlock(
+          txs,
+          merkle.checkpointBlock.checkpoint,
+          msgs,
+          notifications
+        )
+      ),
+      merkle.children,
+      merkle.height
+    )
+  }
+
+  def fetchTransactions(
+    merkleRoot: String
+  )(implicit dao: DAO): Seq[Transaction] = {
+
+    fetch[TransactionCacheData,Transaction](
+      merkleRoot,
+      dao.transactionService,
+      (x: TransactionCacheData) => x.transaction,
+      (s: String) =>
+        DataResolver.resolveTransactions(s,
+          dao.readyPeers.map(p => PeerApiClient(p._1, p._2.client))).map(_.get)
+    )
+  }
+
+  def fetchMessages(merkleRoot: String)(implicit dao: DAO): Seq[ChannelMessage] = {
+    fetch[ChannelMessageMetadata,ChannelMessage](
+      merkleRoot,
+      dao.messageService,
+      (x: ChannelMessageMetadata) => x.channelMessage,
+      (s: String) =>
+        DataResolver.resolveMessages(s,
+          dao.readyPeers.map(p => PeerApiClient(p._1, p._2.client))).map(_.get)
+    )
+  }
+
+  def fetch[T, R](
+    merkleRoot: String,
+    service: MerkleService[T],
+    mapper: T => R,
+    resolver: String => IO[T],
+  )(implicit dao: DAO): Seq[R] = {
+    service.findHashesByMerkleRoot(merkleRoot)
+      .map(
+        _.get
+          .map(hash => service.lookup(hash).map(t => mapper(t.getOrElse(resolver(hash).unsafeRunSync()))))
```

Comment:
You shouldn't perform `unsafeRunSync()` inside for each record, just do it once at the end